Tool install with nuget error should not display the temp project in output

## Steps to reproduce

See related issue #8698.

```
dotnet install tool Microsoft.Dotnet.Watcher.Tools -g
```

## Expected  behavior

The temp project path should not appear in the output by default:

```
$ dotnet install tool Microsoft.Dotnet.Watcher.Tools -g
error NU1212: Invalid project-package combination for Microsoft.DotNet.Watcher.Tools 2.0.0. DotnetToolReference project style can only contain references of the DotnetTool type
The tool package could not be restored.
Tool 'Microsoft.Dotnet.Watcher.Tools' failed to install.
```

## Actual behavior

The temp project path appears in the output:

```
$ dotnet install tool Microsoft.Dotnet.Watcher.Tools -g
/var/folders/rl/m1rv1rxd643gpwspqrbxy76r0000gn/T/w1heen5l.wc5/yh505uu5.u4i.csproj : error NU1212: Invalid project-package combination for Microsoft.DotNet.Watcher.Tools 2.0.0. DotnetToolReference project style can only contain references of the DotnetTool type
The tool package could not be restored.
Tool 'Microsoft.Dotnet.Watcher.Tools' failed to install.
```

## Environment data
```
$ dotnet --info
.NET Command Line Tools (2.1.300-preview2-008266)

Product Information:
 Version:            2.1.300-preview2-008266
 Commit SHA-1 hash:  6fb919f433

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  10.13
 OS Platform: Darwin
 RID:         osx-x64
 Base Path:   /Users/peterhuene/src/cli/bin/2/osx-x64/dotnet/sdk/2.1.300-preview2-008266/

Microsoft .NET Core Shared Framework Host

  Version  : 2.1.0-preview2-26131-06
  Build    : b13a0d5c331f374afd35ded57b9a4b4ab128864c
```

@KathleenDollard Examples with the fix applied for your review: 

```
$ dotnet install tool -g --source https://dotnet.myget.org/F/aspnetcore-dev/api/v3/index.json --version 2.1.0-preview2-\*  dotnet-watch
warning NU1603: tool depends on dotnet-watch (>= 2.1.0-preview2-) but dotnet-watch 2.1.0-preview2- was not found. An approximate best match of dotnet-watch 2.1.0-preview2-28177 was resolved.
If there were no additional instructions, you can type the following command to invoke the tool: dotnet-watch
Tool 'dotnet-watch' (version '2.1.0-preview2-28177') was successfully installed.
```

```
$ dotnet install tool -g --source https://dotnet.myget.org/F/aspnetcore-dev/api/v3/index.json dotnet-watch
error NU1103: Unable to find a stable package dotnet-watch with version
error NU1103:   - Found 264 version(s) in aspnetcore-dev [ Nearest version: 2.1.0-preview1-27934 ]
error NU1103:   - Found 0 version(s) in /Users/peterhuene/src/cli/bin/2/osx-x64/dotnet/sdk/NuGetFallbackFolder
The tool package could not be restored.
Tool 'dotnet-watch' failed to install.
```

```
$ dotnet install tool -g foo
warning NU1701: Package 'Foo 1.0.7.2' was restored using '.NETFramework,Version=v4.6.1' instead of the project target framework '.NETCoreApp,Version=v2.1'. This package may not be fully compatible with your project.
error NU1212: Invalid project-package combination for Foo 1.0.7.2. DotnetToolReference project style can only contain references of the DotnetTool type
The tool package could not be restored.
Tool 'foo' failed to install.
```

Note that this continues to colorize warnings in yellow and errors in red.  The custom filtering is only applied when using the default verbosity level of `quiet`; otherwise, the filtering is disabled and MSBuild inherits the stdout/stderr handles and paths to the temp project will be displayed.

--- a/src/dotnet/ToolPackage/IToolPackageInstaller.cs
+++ b/src/dotnet/ToolPackage/IToolPackageInstaller.cs
@@ -15,6 +15,7 @@ namespace Microsoft.DotNet.ToolPackage
             VersionRange versionRange = null,
             string targetFramework = null,
             FilePath? nugetConfig = null,
+            DirectoryPath? rootConfigDirectory = null,
             string source = null,
             string verbosity = null);
     }

--- a/src/dotnet/ToolPackage/ToolPackageFactory.cs
+++ b/src/dotnet/ToolPackage/ToolPackageFactory.cs
@@ -16,7 +16,7 @@ namespace Microsoft.DotNet.ToolPackage
             IToolPackageStore toolPackageStore = CreateToolPackageStore(nonGlobalLocation);
             var toolPackageInstaller = new ToolPackageInstaller(
                 toolPackageStore,
-                new ProjectRestorer(Reporter.Output));
+                new ProjectRestorer());
 
             return (toolPackageStore, toolPackageInstaller);
         }

--- a/src/dotnet/ToolPackage/ToolPackageInstaller.cs
+++ b/src/dotnet/ToolPackage/ToolPackageInstaller.cs
@@ -119,7 +119,7 @@ namespace Microsoft.DotNet.ToolPackage
         {
             var tempProject = _tempProject ?? new DirectoryPath(Path.GetTempPath())
                 .WithSubDirectories(Path.GetRandomFileName())
-                .WithFile(Path.GetRandomFileName() + ".csproj");
+                .WithFile("restore.csproj");
 
             if (Path.GetExtension(tempProject.Value) != "csproj")
             {

--- a/src/dotnet/ToolPackage/ToolPackageInstaller.cs
+++ b/src/dotnet/ToolPackage/ToolPackageInstaller.cs
@@ -35,6 +35,7 @@ namespace Microsoft.DotNet.ToolPackage
             VersionRange versionRange = null,
             string targetFramework = null,
             FilePath? nugetConfig = null,
+            DirectoryPath? rootConfigDirectory = null,
             string source = null,
             string verbosity = null)
         {
@@ -53,7 +54,8 @@ namespace Microsoft.DotNet.ToolPackage
                             packageId: packageId,
                             versionRange: versionRange,
                             targetFramework: targetFramework ?? BundledTargetFramework.GetTargetFrameworkMoniker(),
-                            restoreDirectory: stageDirectory);
+                            restoreDirectory: stageDirectory,
+                            rootConfigDirectory: rootConfigDirectory);
 
                         try
                         {
@@ -115,7 +117,8 @@ namespace Microsoft.DotNet.ToolPackage
             PackageId packageId,
             VersionRange versionRange,
             string targetFramework,
-            DirectoryPath restoreDirectory)
+            DirectoryPath restoreDirectory,
+            DirectoryPath? rootConfigDirectory)
         {
             var tempProject = _tempProject ?? new DirectoryPath(Path.GetTempPath())
                 .WithSubDirectories(Path.GetRandomFileName())
@@ -135,7 +138,7 @@ namespace Microsoft.DotNet.ToolPackage
                         new XElement("TargetFramework", targetFramework),
                         new XElement("RestorePackagesPath", restoreDirectory.Value),
                         new XElement("RestoreProjectStyle", "DotnetToolReference"), // without it, project cannot reference tool package
-                        new XElement("RestoreRootConfigDirectory", Directory.GetCurrentDirectory()), // config file probing start directory
+                        new XElement("RestoreRootConfigDirectory", rootConfigDirectory?.Value ?? Directory.GetCurrentDirectory()), // config file probing start directory
                         new XElement("DisableImplicitFrameworkReferences", "true"), // no Microsoft.NETCore.App in tool folder
                         new XElement("RestoreFallbackFolders", "clear"), // do not use fallbackfolder, tool package need to be copied to tool folder
                         new XElement("RestoreAdditionalProjectSources", // use fallbackfolder as feed to enable offline

--- a/src/dotnet/commands/dotnet-install/dotnet-install-tool/ProjectRestorer.cs
+++ b/src/dotnet/commands/dotnet-install/dotnet-install-tool/ProjectRestorer.cs
@@ -16,10 +16,14 @@ namespace Microsoft.DotNet.Tools.Install.Tool
     {
         private const string AnyRid = "any";
         private readonly IReporter _reporter;
+        private readonly IReporter _errorReporter;
+        private readonly bool _forceOutputRedirection;
 
         public ProjectRestorer(IReporter reporter = null)
         {
-            _reporter = reporter;
+            _reporter = reporter ?? Reporter.Output;
+            _errorReporter = reporter ?? Reporter.Error;
+            _forceOutputRedirection = reporter != null;
         }
 
         public void Restore(
@@ -56,11 +60,11 @@ namespace Microsoft.DotNet.Tools.Install.Tool
             var command = new DotNetCommandFactory(alwaysRunOutOfProc: true)
                 .Create("restore", argsToPassToRestore);
 
-            if (_reporter != null)
+            if (verbosity == null || _forceOutputRedirection)
             {
                 command = command
-                    .OnOutputLine((line) => _reporter.WriteLine(line))
-                    .OnErrorLine((line) => _reporter.WriteLine(line));
+                    .OnOutputLine(line => WriteLine(_reporter, line, project))
+                    .OnErrorLine(line => WriteLine(_errorReporter, line, project));
             }
 
             var result = command.Execute();
@@ -68,6 +72,29 @@ namespace Microsoft.DotNet.Tools.Install.Tool
             {
                 throw new ToolPackageException(LocalizableStrings.ToolInstallationRestoreFailed);
             }
+        }
+
+        private static void WriteLine(IReporter reporter, string line, FilePath project)
+        {
+            line = line ?? "";
+
+            // Remove the temp project prefix if present
+            if (line.StartsWith($"{project.Value} : ", StringComparison.OrdinalIgnoreCase))
+            {
+                line = line.Substring(project.Value.Length + 3);
+            }
+
+            // Note: MSBuild intentionally does not localize "warning" and "error" for diagnostic messages
+            if (line.StartsWith("warning ", StringComparison.Ordinal))
+            {
+                line = line.Yellow();
+            }
+            else if (line.StartsWith("error ", StringComparison.Ordinal))
+            {
+                line = line.Red();
+            }
+
+            reporter.WriteLine(line);
         }
     }
 }

--- a/test/Microsoft.DotNet.ToolPackage.Tests/ToolPackageInstallerTests.cs
+++ b/test/Microsoft.DotNet.ToolPackage.Tests/ToolPackageInstallerTests.cs
@@ -154,21 +154,17 @@ namespace Microsoft.DotNet.ToolPackage.Tests
         }
 
         [Fact]
-        public void GivenAConfigFileInCurrentDirectoryPackageInstallSucceeds()
+        public void GivenAConfigFileRootDirectoryPackageInstallSucceeds()
         {
             var nugetConfigPath = WriteNugetConfigFileToPointToTheFeed();
 
             var (store, installer, reporter, fileSystem) = Setup(useMock: false);
 
-            /*
-             * In test, we don't want NuGet to keep look up, so we point current directory to nugetconfig.
-             */
-            Directory.SetCurrentDirectory(nugetConfigPath.GetDirectoryPath().Value);
-
             var package = installer.InstallPackage(
                 packageId: TestPackageId,
                 versionRange: VersionRange.Parse(TestPackageVersion),
-                targetFramework: _testTargetframework);
+                targetFramework: _testTargetframework,
+                rootConfigDirectory: nugetConfigPath.GetDirectoryPath());
 
             AssertPackageInstall(reporter, fileSystem, package, store);
 
@@ -641,7 +637,7 @@ namespace Microsoft.DotNet.ToolPackage.Tests
             FilePath? tempProject = null,
             DirectoryPath? offlineFeed = null)
         {
-            var root = new DirectoryPath(Path.Combine(Directory.GetCurrentDirectory(), Path.GetRandomFileName()));
+            var root = new DirectoryPath(Path.Combine(Path.GetFullPath(TempRoot.Root), Path.GetRandomFileName()));
             var reporter = new BufferedReporter();
 
             IFileSystem fileSystem;

--- a/test/Microsoft.DotNet.ToolPackage.Tests/ToolPackageInstallerTests.cs
+++ b/test/Microsoft.DotNet.ToolPackage.Tests/ToolPackageInstallerTests.cs
@@ -498,6 +498,32 @@ namespace Microsoft.DotNet.ToolPackage.Tests
             package.Uninstall();
         }
 
+        [Fact]
+        public void GivenANuGetDiagnosticMessageItShouldNotContainTheTempProject()
+        {
+            var nugetConfigPath = WriteNugetConfigFileToPointToTheFeed();
+            var tempProject = GetUniqueTempProjectPathEachTest();
+
+            var (store, installer, reporter, fileSystem) = Setup(
+                useMock: false,
+                tempProject: tempProject);
+
+            var package = installer.InstallPackage(
+                packageId: TestPackageId,
+                versionRange: VersionRange.Parse("1.0.*"),
+                targetFramework: _testTargetframework,
+                nugetConfig: nugetConfigPath);
+
+            reporter.Lines.Should().NotBeEmpty();
+            reporter.Lines.Should().Contain(l => l.Contains("warning"));
+            reporter.Lines.Should().NotContain(l => l.Contains(tempProject.Value));
+            reporter.Lines.Clear();
+
+            AssertPackageInstall(reporter, fileSystem, package, store);
+
+            package.Uninstall();
+        }
+
         private static void AssertPackageInstall(
             BufferedReporter reporter,
             IFileSystem fileSystem,

--- a/test/Microsoft.DotNet.Tools.Tests.ComponentMocks/ToolPackageInstallerMock.cs
+++ b/test/Microsoft.DotNet.Tools.Tests.ComponentMocks/ToolPackageInstallerMock.cs
@@ -40,6 +40,7 @@ namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
             VersionRange versionRange = null,
             string targetFramework = null,
             FilePath? nugetConfig = null,
+            DirectoryPath? rootConfigDirectory = null,
             string source = null,
             string verbosity = null)
         {

